src/factor_graph.cc

```diff
@@ -84,25 +84,32 @@ FactorGraph::~FactorGraph() {
   fast_alloc_free(values.release());
 }
 
+// Construct index using multiple threads
+// 6 sec instead of 30 sec for a 270M-factor graph
+// std::vector<std::tuple<size_t, size_t, size_t, size_t>> params;
 void FactorGraph::construct_index() {
   size_t total = size.num_variables;
-
-  // large graph, multi thread
-  // 6 sec instead of 30 sec for a 270M-factor graph
-  std::vector<std::tuple<size_t, size_t, size_t, size_t>> params;
   size_t cores = sysconf(_SC_NPROCESSORS_CONF);
   size_t increment = total / cores;
 
+  // small graph, single thread
+  if (total < 1000) {
+    increment = total;
+  }
+
   size_t milestone = 0;
   size_t num_values = 0, num_factors = 0;
 
+  std::vector<std::function<void()>> tasks;
   for (size_t i = 0; i < size.num_variables; ++i) {
     if (i == milestone) {
       milestone += increment;
       if (milestone > total) {
         milestone = total;
       }
-      params.push_back(std::make_tuple(i, milestone, num_values, num_factors));
+      tasks.push_back([this, i, milestone, num_values, num_factors]() {
```

Comment:
I think `[=, this]() { ... }` will probably do the same job without repetition, but will leave it up to you.